### PR TITLE
Cloud awareness, continued

### PR DESCRIPTION
**Pull request by nilsanderselde**
### Which issue this PR addresses:

Partially fixes [8763812](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/8763812)

### What this PR does / why we need it:

- Use correct domains for current Azure cloud environment

### Test plan for issue:

Run existing unit and E2E tests.

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

**Comment by nilsanderselde** [comment]
Would need to set `AZURE_ENVIRONMENT` env var in E2E for current authorizer changes to work

File: deploy/rp-production.json
Comment (nilsanderselde) [reply deploy/rp-production.json:223]: This change belongs to a different PR to be merged first: see https://github.com/Azure/ARO-RP/pull/1101#discussion_r513878288. Basically, it doesn't matter, because the `.azurecr.io` currently gets trimmed anyway.

File: pkg/cluster/deployresources_resources.go
Comment (jim-minter) [inline pkg/cluster/deployresources_resources.go:97]: as part of these changes the oc argument can be removed - please do so.  Check in case there are other arguments that can be removed?

File: pkg/util/clientauthorizer/arm.go
Comment (jim-minter) [inline pkg/util/clientauthorizer/arm.go:102]: store instancemetadata.InstanceMetadata as a field inside the arm type and read the environment from there

File: .pipelines/templates/template-prod-e2e-steps.yml
Comment (jim-minter) [inline .pipelines/templates/template-prod-e2e-steps.yml:14]: I'm not sure I'm mentally ready to merge the .pipelines or Makefile changes in this PR.  You can try to convince me though.

My concerns are:
1) we're in the middle of trying to simplify pipelines, but this adds complexity which isn't actually in use yet
2) we don't yet own the ACR and storage account names in govcloud, if someone else has them this logic would have  to change
3) I would like to have a process to promote built images from int->prod but we don't yet know how we'll want to architect that for govt cloud.  There's no guarantee that the changes proposed here will take us in the right direction.

We'll need to do these things, I just think it's a bit early still.

File: cmd/aro/mirror.go
Comment (jim-minter) [inline cmd/aro/mirror.go:46]: I think that mirror should consume an env.Core rather than doing this

File: pkg/deploy/deploy.go
Comment (jim-minter) [inline pkg/deploy/deploy.go:73]: I think this should also use env.Core - is that possible?

File: pkg/util/azureenvironment/environment.go
Comment (jim-minter) [inline pkg/util/azureenvironment/environment.go:14]: if we use env.Core everywhere, hopefully this goes away again?

**Review by jim-minter — CHANGES_REQUESTED** [review]

File: pkg/deploy/deploy.go
Comment (nilsanderselde) [reply pkg/deploy/deploy.go:73]: I think so - I thought it might be more than was needed, but it would simplify this

File: .pipelines/templates/template-prod-e2e-steps.yml
Comment (nilsanderselde) [reply .pipelines/templates/template-prod-e2e-steps.yml:14]: OK, sounds good. I'll pull these commits out.

File: pkg/cluster/deployresources_resources.go
Comment (nilsanderselde) [reply pkg/cluster/deployresources_resources.go]: Done

File: pkg/util/azureenvironment/environment.go
Comment (nilsanderselde) [reply pkg/util/azureenvironment/environment.go]: it's gone

File: pkg/util/azureclient/mgmt/redhatopenshift/openshiftclusters.go
Comment (nilsanderselde) [reply pkg/util/azureclient/mgmt/redhatopenshift/openshiftclusters.go]: done. it's the last commit in this PR, because it's coupled with a preceding commit.

File: pkg/deploy/generator/templates.go
Comment (nilsanderselde) [reply pkg/deploy/generator/templates.go]: changed

File: pkg/deploy/deploy.go
Comment (nilsanderselde) [reply pkg/deploy/deploy.go]: done

File: cmd/aro/mirror.go
Comment (nilsanderselde) [reply cmd/aro/mirror.go]: done

File: pkg/util/clientauthorizer/arm.go
Comment (nilsanderselde) [reply pkg/util/clientauthorizer/arm.go]: Done.

To make sure the `arm_test.go` tests still worked, I added a function, `instancemetadata.NewFake`, which doesn't require environment variables to be set. I didn't put it in its own subpackage because `instanceMetadata` is unexported. Open to suggestions about this.

**Comment by nilsanderselde** [comment]
@jim-minter Can you please take another look?

File: pkg/backend/openshiftcluster/openshiftcluster.go
Comment (mjudeikis) [inline pkg/backend/openshiftcluster/openshiftcluster.go:54]: We still have a number of other places where we use `_env` as a pattern. I would like to keep it as it is with `_env` considering the package is names same `env`. Which is causing confusion and overload of the same word/term.

File: pkg/util/instancemetadata/fake.go
Comment (ihcsim) [inline pkg/util/instancemetadata/fake.go:29]: This code looks identical to that found in [`pkg/util/instancemetadata/dev.go`](https://github.com/Azure/ARO-RP/blob/d2e6fcbcba9900740e76efe0a3b508d2bac5fc87/pkg/util/instancemetadata/dev.go#L25-L40). Any reasons why they can't be DRY up?

Not sure how this unit test is set up, but generally, I try to avoid introducing external dependencies like env vars in unit tests, to avoid potential flakiness. Where are these env vars being set?

File: cmd/aro/mirror.go
Comment (ihcsim) [inline cmd/aro/mirror.go:51]: We should check the returned value of `azure.ContainerRegistryDNSSuffix` to make sure it is a legit DNS suffix. In [some cases](https://github.com/Azure/go-autorest/blob/f7ec17d62576b689bfc665ead655066c00c9d363/autorest/azure/environments.go#L213), the Azure SDK will return the`Not Available` literal. I am not 100% sure if we can just replace that with the `azure.PublicCloud.ContainerRegistryDNSSuffix`.

**Review by ihcsim — COMMENTED** [review]
Code change LGTM. See below for comments. LMKWYT.

I will defer to @jim-minter and @mjudeikis for the final approval since they have more context.

Also, in general, I find it challenging to correctly and efficiently review big PRs. One thing we can consider doing in the future is breaking up the PRs into smaller logical pieces. In this case, we might be able to perform the refactor tasks in `pkg/cluster/deployresources*.go` and the name changes in a first PR. Then follow up by a separate PR with the actual cloud awareness implementation.

File: pkg/backend/openshiftcluster/openshiftcluster.go
Comment (nilsanderselde) [reply pkg/backend/openshiftcluster/openshiftcluster.go:54]: Discussed this - going to stay with current convention of only using `_env` as a name if necessary.

File: pkg/util/instancemetadata/fake.go
Comment (nilsanderselde) [reply pkg/util/instancemetadata/fake.go:29]: I will hold off on refactoring to avoid repetition for now because this entire approach might be ill-advised.

To explain the purpose of this, in a couple unit tests, there is a new dependency on `InstanceMetadata`. Currently, the simplest way of making one of these from outside the `instancemetadata` package requires that env vars be set, which is not the case when running unit tests.

File: cmd/aro/mirror.go
Comment (nilsanderselde) [reply cmd/aro/mirror.go:51]: The immediate purpose of this work is to prepare to have ARO in US Government Cloud. Adding validation for the event that German Cloud is the current cloud environment seems premature.

@jim-minter, what do you think?

**Comment by nilsanderselde** [comment]
> Also, in general, I find it challenging to correctly and efficiently review big PRs. One thing we can consider doing in the future is breaking up the PRs into smaller logical pieces.

I agree. This PR is the second part of a story which was already broken up, but now it has gotten big itself. I would suggest reviewing the code changes one commit at a time for this PR, as I have intended them to be cohesive chunks of work.

File: cmd/aro/mirror.go
Comment (ihcsim) [reply cmd/aro/mirror.go:51]: I was hoping we can just do a regex validation to determine if it's a legit DNS name or not, regardless of which cloud.

**Comment by ihcsim** [comment]
> I would suggest reviewing the code changes one commit at a time for this PR

Do comments to the commits get folded into the PR? Last I checked, I think commit comments are attached to specific commits only; not visible in the PR.

File: pkg/util/instancemetadata/fake.go
Comment (nilsanderselde) [reply pkg/util/instancemetadata/fake.go]: Removing this and using `MockInstanceMetadata` as should have been the obvious move :facepalm:

File: cmd/aro/mirror.go
Comment (jim-minter) [reply cmd/aro/mirror.go:51]: How about adding a validator in env.NewCore that validates that the Environment is a recognised one?

**Comment by nilsanderselde** [comment]
@ihcsim @jim-minter I've pulled two commits out of this branch into their own branches to make it easier to review
- https://github.com/Azure/ARO-RP/pull/1192
- https://github.com/Azure/ARO-RP/pull/1193

File: pkg/util/rpauthorizer/rpauthorizer.go
Comment (jim-minter) [inline pkg/util/rpauthorizer/rpauthorizer.go:27]: don't do this - pass in the instancemetadata via rpauthorizer.New()

**Review by jim-minter — COMMENTED** [review]
Awesome work @nilsanderselde.  Just a couple of nits and we'll merge.  Don't add anything more to this PR otherwise please.

**Comment by jim-minter** [comment]
@nilsanderselde MJ and I have reviewed without needing #1192 or #1193, those can probably be closed I guess and we'll merge this one?

File: cmd/aro/mirror.go
Comment (nilsanderselde) [reply cmd/aro/mirror.go:51]: sure

File: pkg/util/rpauthorizer/rpauthorizer.go
Comment (nilsanderselde) [reply pkg/util/rpauthorizer/rpauthorizer.go]: changed

File: cmd/aro/mirror.go
Comment (nilsanderselde) [reply cmd/aro/mirror.go:51]: Created [8872444](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/8872444) for follow-up work

**Comment by nilsanderselde** [comment]
> @nilsanderselde MJ and I have reviewed without needing #1192 or #1193, those can probably be closed I guess and we'll merge this one?

(replying again here for visibility) They are necessary; they contain commits which are no longer in this branch.

**Comment by jim-minter** [comment]
Closing for https://github.com/Azure/ARO-RP/pull/1194

**Pull request closed without merging**
